meetings/2021-04-28.md

```diff
@@ -0,0 +1,358 @@
+4. Merged 201
+
+
+## ACP apply policy algorithm
+
+**Henry**: Last week, we discussed a PR I made to 
+[the wac-acp-diff story](https://github.com/solid/authorization-panel/blob/main/proposals/wac-acp-diff-story.md) which we just committed.
+We looked at how the ACL ontology allows a behavior similar to ACP by creating an inferred  `:authorizes` relation 
+defined as:
+
+```Turtle
+:authorizes owl:inverseOf [
+    owl:propertyChainAxiom(:accessControl acl:accessTo)
+]
+```
+
+This relation, which is already implicit in ACL, as it can be defined purely in terms of those concepts, allows one to get similar behavior to the one put forward as the distinguishing feature of ACP.
+
+We then looked at the ShEx examples and showed that with a very simple inferencing step -- adding one
+triple to the graph after following an `:accessControl` `Link` header relation -- one could even apply that WAC ShEx to that graph and get clients written to the old ACL ontology to continue working.
+
+But the question then was what do the `allOf`, `anyOf`, `noneOf` and `apply*` relations bring to the picture. How do they work?
+
+We were about to start on that when time ran out.
+We had just pasted this code into the chat.
+
+```typescript
+/**
+ * # The apply policy algorithm
+ *
+ * A policy is applied if:
+ * - At least one matcher it uses matches the given context; and
+ * - All the conditions it defines are satisfied.
+ *
+ * Policy conditions behave like intersection, union and exclusion operators. In
+ * order to be satisfied:
+ * - allOf requires all of its matchers to match the given context;
+ * - anyOf requires one of its matchers to match the given context;
+ * - noneOf requires none of its matchers to match the given context.
+ *
+ * Note: Given that the noneOf condition excludes matches, a policy without a
+ * satisfied allOf or anyOf condition never applies.
+ *
+ * @param policy The policy to evaluate
+ * @param context The context in which the policy is evaluated
+ * @returns True if the policy applies, false otherwise
+ */
+export function applyPolicy(policy: IPolicy, context: IContext): boolean {
+  const matchContext = (matcher: IMatcher) => {
+    return matcher.match(context);
+  };
+
+  // A policy without positive match doesn't apply
+  if (policy.allOf.length + policy.anyOf.length === 0) {
+    return false;
+  }
+
+  return (
+    (policy.allOf.length === 0 || policy.allOf.every(matchContext)) &&
+    (policy.anyOf.length === 0 || policy.anyOf.some(matchContext)) &&
+    (policy.noneOf.length === 0 || !policy.noneOf.some(matchContext))
+  );
+}
+```
+
+**Henry**: before we look at that code, can we look at some simple RDF examples of a policy using that vocabulary?
+  
+**Mattieu** 
+
+Yes, I can write it out here.
+
+```Turtle
+prefix ex: <https://example.com/>
+
+ex:PolicyX
+    acp:allow acl:Read ; # would only having acp:allow acl:Read imply anything other than Read? Why would acp:deny be necessary?
```

Comment:
I am not sure who wrote that comment btw. I think it may have been @csarven . (Is that right?) 
We should put the name of who wrote it in there, and then we can add your answer here too. Though it could be a discussion that needs to be pursued at more length I think.

Reply:
That makes sense. Waiting for Sarven's answer before suggesting a change.

Reply:
I checked that the [HackMd from Wednesday](https://hackmd.io/mFgvpYMZR7Co_DIDgUaNwA) attributes that sentence to Sarven.

Reply:
```suggestion
    acp:allow acl:Read ; # Sarven: would only having acp:allow acl:Read imply anything other than Read? Why would acp:deny be necessary?
```

Reply:
Super, I added my comment below and suggest attributing to Sarven as above.